src/Common/src/TypeSystem/IL/ILStackHelper.cs

```diff
@@ -0,0 +1,479 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+using Internal.TypeSystem;
+
+namespace Internal.IL
+{
+    public static class ILStackHelper
+    {
+        /// <summary>
+        /// Computes the maximum number of items that can be pushed onto the CIL evaluation stack.
+        /// </summary>
+        public static int ComputeMaxStack(this MethodIL methodIL)
+        {
+            return methodIL.ComputeMaxStack(assert: false);
+        }
+
+        /// <summary>
+        /// Validates that the CIL evaluation stack is properly balanced.
+        /// </summary>
+        [Conditional("DEBUG")]
+        public static void CheckStackBalance(this MethodIL methodIL)
+        {
+            methodIL.ComputeMaxStack(assert: true);
+        }
+
+        private static int ComputeMaxStack(this MethodIL methodIL, bool assert = false)
```

Comment:
Why not assert all the time?

Reply:
For some reason I thought it would be bad if we were hitting the asserts in the debugger visualizer for `MethodIL` (the IL disassembler), but the debugger seems to be eating the asserts there. So no reason. Removed.